contracts/oracles/v2/adapters/balancer/BalancerV2PriceOracle.sol

```diff
@@ -0,0 +1,169 @@
+            .getTimeWeightedAverage(query);
+        uint256 tokenPairPrice = prices[0];
+
+        int256 denominatedTokenPrice = AggregatorV3Interface(denominatedOracles[_token])
+            .latestAnswer();
+        uint8 denominatedTokenPriceDecimals = AggregatorV3Interface(
+            denominatedOracles[_token]
+        ).decimals();
+        bytes32 poolId = IBPoolV2(tokenPools[_token]).getPoolId();
+        (IERC20[] memory tokens, , ) = IBVaultV2(vault).getPoolTokens(poolId);
+
+        if (_token == address(tokens[0])) {
+            price =
+                (uint256(denominatedTokenPrice) * (10**36)) /
+                tokenPairPrice /
+                (10**denominatedTokenPriceDecimals);
+        } else if (_token == address(tokens[1])) {
+            price =
+                (uint256(denominatedTokenPrice) * tokenPairPrice) /
+                (10**denominatedTokenPriceDecimals);
+        }
+    }
+
+    /**
+     * @dev returns the current "unsafe" price that can be easily manipulated
+     */
+    function getCurrentPrice(address _token) external view returns (uint256 price) {
+        require(
+            tokenPools[_token] != address(0) && denominatedOracles[_token] != address(0),
+            "UNSUPPORTED"
+        );
+
+        int256 denominatedTokenPrice = AggregatorV3Interface(denominatedOracles[_token])
+            .latestAnswer();
+        uint8 denominatedTokenPriceDecimals = AggregatorV3Interface(
+            denominatedOracles[_token]
+        ).decimals();
+
+        bytes32 poolId = IBPoolV2(tokenPools[_token]).getPoolId();
+        uint256[] memory weights = IBPoolV2(tokenPools[_token]).getNormalizedWeights();
+        (IERC20[] memory tokens, uint256[] memory balances, ) = IBVaultV2(vault)
+            .getPoolTokens(poolId);
+
+        if (_token == address(tokens[0])) {
+            // price = balance1 / balance0 * weight0 / weight1 * usdPrice1
+            // in denominated token price decimals
+            uint256 assetValue = (balances[1] * uint256(denominatedTokenPrice)) /
+                (10**tokens[1].decimals());
+            // in denominated token price decimals
+            uint256 tokenPrice = (assetValue * weights[0] * (10**tokens[0].decimals())) /
+                weights[1] /
+                balances[0];
+
+            price = (tokenPrice * (10**18)) / (10**denominatedTokenPriceDecimals);
+        } else if (_token == address(tokens[1])) {
+            // price = balance0 / balance1 * weight1 / weight0 * usdPrice0
+            // in denominated token price decimals
+            uint256 assetValue = (balances[0] * uint256(denominatedTokenPrice)) /
+                (10**tokens[0].decimals());
+            // in denominated token price decimals
+            uint256 tokenPrice = (assetValue * weights[1] * (10**tokens[1].decimals())) /
+                weights[0] /
+                balances[1];
+
+            price = (tokenPrice * (10**18)) / (10**denominatedTokenPriceDecimals);
+        }
```

Comment:
Refactor this to extract the math calculating `assetValue`, `tokenPrice` and `price` to a function which returns `price`